hy/contrib/hy_repr.hy

```diff
@@ -4,95 +4,161 @@
+    (+ (get syntax (first x)) (hy-repr (second x)))
+    (+ "(" (-cat x) ")"))))
+
+(hy-repr-register HySymbol str)
+(hy-repr-register [str-type bytes-type HyKeyword] (fn [x]
+  (if (and (instance? str-type x) (.startswith x HyKeyword.PREFIX))
+    (return (cut x 1)))
+  (setv r (.lstrip (-base-repr x) "ub"))
+  (+
+    (if (instance? bytes-type x) "b" "")
+    (if (.startswith "\"" r)
+      ; If Python's built-in repr produced a double-quoted string, use
+      ; that.
+      r
+      ; Otherwise, we have a single-quoted string, which isn't valid Hy, so
+      ; convert it.
+      (+ "\"" (.replace (cut r 1 -1) "\"" "\\\"") "\"")))))
+(hy-repr-register bool str)
+(if (not PY3) (hy-repr-register int (fn [x]
+  (.format "(int {})" (-base-repr x)))))
+(if (not PY3) (hy-repr-register long_type (fn [x]
+  (.rstrip (-base-repr x) "L"))))
+(hy-repr-register float (fn [x]
+  (if
+    (isnan x)  "NaN"
+    (= x Inf)  "Inf"
+    (= x -Inf) "-Inf"
+               (-base-repr x))))
+(hy-repr-register complex (fn [x]
+  (.replace (.replace (.strip (-base-repr x) "()") "inf" "Inf") "nan" "NaN")))
+(hy-repr-register fraction (fn [x]
+  (.format "{}/{}" (hy-repr x.numerator) (hy-repr x.denominator))))
+
+(setv -matchobject-type (type (re.match "" "")))
+(hy-repr-register -matchobject-type (fn [x]
+  (.format "<{}.{} object; :span {} :match {}>"
+    -matchobject-type.__module__
+    -matchobject-type.__name__
+    (hy-repr (.span x))
+    (hy-repr (.group x 0)))))
+
+(hy-repr-register datetime.datetime (fn [x]
+  (.format "(datetime.datetime {}{})"
+    (.strftime x "%Y %-m %-d %-H %-M %-S")
+    (-repr-time-innards x))))
+(hy-repr-register datetime.date (fn [x]
+  (.strftime x "(datetime.date %Y %-m %-d)")))
+(hy-repr-register datetime.time (fn [x]
+  (.format "(datetime.time {}{})"
+    (.strftime x "%-H %-M %-S")
+    (-repr-time-innards x))))
+(defn -repr-time-innards [x]
+  (.rstrip (+ " " (.join " " (filter identity [
+    (if x.microsecond (str-type x.microsecond))
+    (if (not (none? x.tzinfo)) (+ ":tzinfo " (hy-repr x.tzinfo)))
+    (if (and PY36 (!= x.fold 0)) (+ ":fold " (hy-repr x.fold)))])))))
+
+(hy-repr-register collections.Counter (fn [x]
+  (.format "(Counter {})"
+    (hy-repr (dict x)))))
+(hy-repr-register collections.defaultdict (fn [x]
+  (.format "(defaultdict {} {})"
+    (hy-repr x.default-factory)
+    (hy-repr (dict x)))))
+
+(for [[types fmt] (partition [
+    list "[...]"
+    [set HySet] "#{...}"
+    frozenset "(frozenset #{...})"
+    dict-keys "(dict-keys [...])"
+    dict-values "(dict-values [...])"
+    dict-items "(dict-items [...])"])]
+  (defn mkrepr [fmt]
+    (fn [x] (.replace fmt "..." (-cat x) 1)))
+  (hy-repr-register types :placeholder fmt (mkrepr fmt)))
+
+(defn -cat [obj]
+  (.join " " (map hy-repr obj)))
 
-(defn base-repr [x]
+(defn -base-repr [x]
   (unless (instance? HyObject x)
     (return (repr x)))
   ; Call (.repr x) using the first class of x that doesn't inherit from
```

Comment:
Why not check for `__hy_repr__` as well? I don't think defining that as a protocol was a bad idea, tbh.

Reply:
It seems an unnecessary complication. It would bring up issues such as what code gets run when both a `__hy_repr__` and a registered function apply.

Reply:
What if you just threw an exception registration in there's a `__hy_repr__`?

Not a strong opinion either way, but `__hy_repr__` feels like it fits into the Python way a bit better, and its easier for first class support, leaving registration for everything else.

Reply:
I doubt it would be that simple—for example, there's nothing stopping somebody from monkey-patching a `__hy_repr__` onto a class after a function for it has been registered.

A priori, `__hy_repr__` does make more sense than registered functions, but Hy is a second-class citizen in the Python world, and so a Hy programmer is likely to want a Hy representation of a type from a package written by somebody who doesn't care about Hy.

Reply:
If you're crazy, by the way, you can always register a function for `object` that looks for `__hy_repr__`. ;)

Reply:
Good points. I'm comfortable with registration, so don't let me stop this...